onnxruntime/python/session.py

```diff
@@ -303,3 +330,80 @@ def clear_binding_inputs(self):
 
     def clear_binding_outputs(self):
         self._iobinding.clear_binding_outputs()
+
+
+class OrtValue:
+    '''
+    This class provides APIs to construct and deal with OrtValues.
+    '''
+    def __init__(self, ortvalue):
+        if isinstance(ortvalue, C.OrtValue):
+            self._ortvalue = ortvalue
+        else:
+            # An end user won't hit this error
+            raise ValueError("`Provided ortvalue` needs to be of type " +
+                             "`onnxruntime.capi.onnxruntime_pybind11_state.OrtValue`")
+
+    @staticmethod
+    def tensor_from_numpy(numpy_obj, device_type='cpu', device_id=0):
+        '''
+        Factory method to construct an OrtValue (which holds a Tensor) from a given Numpy object
+        A copy of the data in the Numpy object is held by the OrtValue
+        :param numpy_obj: The Numpy object to construct the Tensor from
+        :param device_type: e.g. cpu, cuda, cpu by default
+        :param device_id: device id, e.g. 0
+        '''
+        return OrtValue(C.OrtValue.tensor_from_numpy(numpy_obj, C.OrtDevice(get_ort_device_type(device_type),
+                        C.OrtDevice.default_memory(), device_id)))
+
+    @staticmethod
+    def tensor_from_shape_and_type(shape=None, element_type=None, device_type='cpu', device_id=0):
+        '''
+        Factory method to construct an OrtValue (which holds a Tensor) from given shape and element_type
```

Comment:
>Factory [](start = 8, length = 7)

It does not say anything about the data that it does/does not contain and what this kind of OrtValue is usable for.

Reply:
The memory is left uninitalized initially. This kind of OrtValue may be used for binding an output and for ORT to ultimately write results into. I ll add a comment.

Reply:
What kind of allocator is used to allocate?

---
In reply to: [493037426](https://github.com/microsoft/onnxruntime/pull/5248#discussion_r493037426) [](ancestors = 493037426)

Reply:
ORT's own device specific allocator - (the default CPU allocator / CUDA allocator based on the user provided device argument). Its entire life-cycle is managed by ORT and is bound to the life cycle of the OrtValue instance itself.